staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -474,11 +474,13 @@ func (c *Cacher) startCaching(stopChannel <-chan struct{}) {
 		successfulList = true
 		c.ready.set(true)
 		klog.V(1).Infof("cacher (%v): initialized", c.groupResource.String())
+		metrics.WatchCacheReady.WithLabelValues(c.groupResource.String()).Set(1)
 		metrics.WatchCacheInitializations.WithLabelValues(c.groupResource.String()).Inc()
```

Comment:
I am in favor of adding a new metric to track the state of the cache. 

I think we should consider removing `WatchCacheInitializations`, though I am not sure about the current deprecation policy. 

By using functions like `changes()`, we could count the number of times the cache was initialized, which would be equivalent to the existing counter metrics. @dgrisonnet , please correct me if I am wrong here.

Reply:
`WatchCacheInitializations` can be a bit more precise than using functions on the new metric since its count isn't impacted by the scrape interval or scrape failures. If you solely rely on the new metric you might miss some cache initialization but maybe that level correctness isn't very necessary here.

This is the deprecation policy that we currently follow for metrics: https://kubernetes.io/docs/reference/using-api/deprecation-policy/#deprecating-a-metric. So you could just remove the `WatchCacheInitializations` metric if it makes sense here.

Reply:
It looks like `watchCacheInitializations` was added a long time ago in https://github.com/kubernetes/kubernetes/pull/106737. @wojtek-t since you originally added it, what do you think?

Reply:
WatchCacheInitializations is a better metric for this because it guarantees you are not dependent on sampling rate. If watch cache is unavailable for 5s every 1 minute you might not notice it via WatchCacheReady metric.